torch/_inductor/codecache.py

```diff
@@ -776,6 +799,32 @@ def __init__(
                 else:
                     self.fx_kwargs[k] = v
 
+        from torch._higher_order_ops.triton_kernel_wrap import (
+            TritonKernelWrapperFunctional,
+            TritonKernelWrapperMutation,
+        )
+
+        # Node meta will not be part of gm's reduce function, so lets remember
+        # the kernel source code separately
+        self.user_defined_triton_source: List[Any] = []
+        if gm is not None:
+            for module in gm.modules():
+                if not isinstance(module, torch.fx.GraphModule):
+                    continue
+                for node in module.graph.nodes:
+                    if isinstance(
+                        node.target,
+                        (TritonKernelWrapperFunctional, TritonKernelWrapperMutation),
+                    ):
+                        data = node.meta.get(
+                            "user_defined_triton_kernel_source_and_constant_args", None
+                        )
```

Comment:
@oulgen We've got some options:
- toss the triton kernel source into node.meta
- toss the triton kernel source into graph.meta
- toss the triton kernel source as a string input to the node (we don't need to read from the source if the triton kernel is still alive; we can also probably develop some mechanism for this to print in a sane way).

Ditto for the constant args (in option 3, if the constant args are not fx-graphable, then we could just say they're not cacheable).

One question is: how much do you care about the triton kernel export serializability work? There could be one solution that works for both triton kernel export serializability and the caching work.

Reply:
I kind of like storing the triton kernel source in the graph.meta -- the claim here is that we are essentially going to move the side table to be in the graph itself.

Reply:
I think what you're suggesting is a big refactor to get rid of side table all together which likely requires us to eval the source code for inductor so that we have a function pointer again, or store all the kernel properties ahead of time. For caching purposes only, i wanted to start with just the source since we dont need to restore anything but only use it as a key.

The problem with storing in graph.meta is if someone dropped the meta or removed one in an opt pass, we can no longer assert that we have all of them, as opposed to the current assertion. wdyt?

Reply:
My suggestion is more: I'm worried that we're going to end up with two different implementations, one for caching and one for serialization, or that when we implement serialization we will end up re-implementing caching. Since the timeline for serialization is also soon (@ydwu4's planning on working on it in the next couple of weeks?) I wonder if it makes sense to slow down and figure out how to co-design these. Feel free to make an argument that the code in this PR is easy to rip out, for example.

I'm not suggesting that we should eval the source code for inductor -- I can see how that would be undesirable.

> The problem with storing in graph.meta is if someone dropped the meta or removed one in an opt pass, we can no longer assert that we have all of them, as opposed to the current assertion. wdyt?

In either (1) or (2) it seems like we need to build some assertions into the system. For node.meta we would need to build assertions that for all triton_kernel_* nodes, the node.meta must be populated. For graph.meta, we would need to build the assertion that if the graph has any triton_kernel_* nodes, the triton kernels they are referring to have their sources in graph.meta. A benefit of building into graph.meta is that we can deduplicate nodes (idk if that actually matters for things like cache key computation)

Reply:
@zou3519 I agree eval is undesirable (e.g., as we'd need to collect all transitive dependencies, relevant imports etc.; I think, we rely on `Autotuner` and `JITFunction` instances in multiple places, so we'd need to be able to reliably recreate them from the source).

> I'm not suggesting that we should eval the source code for inductor -- I can see how that would be undesirable.

Given just the source, how can we then reconstruct the kernel instance w/o eval?

Reply:
Did we consider serializing the side table instead of the kernel source? Serialized Triton kernel source (in Python) makes little sense w/o Python runtime, and with it we could also deserialize the kernel instances in the side table?

Reply:
@zou3519 yep your points make sense, but I do worry that this requires a larger refactor around user defined triton kernels, and i'm primarily trying to unblock ads models as their compile time reduction is blocked on lack of user defined triton kernel caching. Would you be willing to proceed with this and later rip this out or move the assertions to graph? Moving the assertions to graph instead of node is trickier which is why i chose this for the quick-and-dirty solution. If you think this is too much tech debt, I'm happy to accept that.

Reply:
@aakhundov Deserialization is the hard question here, esp if we want to do that without eval, by storing a ton of metadata, and since I didnt need it for hashing, I tried to avoid it. But I dont think serde of the side table is any harder than serde of a single kernel pointer, so i think it boils down to the same problem.

Reply:
>  i'm primarily trying to unblock ads models as their compile time reduction is blocked on lack of user defined triton kernel caching. Would you be willing to proceed with this and later rip this out or move the assertions to graph?

Yes I'm happy to proceed with the current design, esp bc there is a pressing need for this right now